runtime/misc.c

```diff
@@ -36,18 +36,26 @@ _Atomic caml_timing_hook caml_finalise_begin_hook = (caml_timing_hook)NULL;
 _Atomic caml_timing_hook caml_finalise_end_hook = (caml_timing_hook)NULL;
 
 #ifdef DEBUG
+void caml_print_loc(char_os *file_os, int line) {
+  char* file = caml_stat_strdup_of_os(file_os);
+  fprintf(stderr, "[%02d] file %s; line %d ### ",
+          (Caml_state_opt != NULL) ? Caml_state_opt->id : -1, file, line);
+  caml_stat_free(file);
+}
 
 void caml_failed_assert (char * expr, char_os * file_os, int line)
 {
-  char* file = caml_stat_strdup_of_os(file_os);
-  fprintf(stderr, "[%02d] file %s; line %d ### Assertion failed: %s\n",
-          (Caml_state_opt != NULL) ? Caml_state_opt->id : -1, file, line, expr);
+  caml_print_loc(file_os, line);
+  fprintf(stderr, "Assertion failed: %s\n", expr);
   fflush(stderr);
-  caml_stat_free(file);
-#if __has_builtin(__builtin_trap) || defined(__GNUC__)
-  __builtin_trap();
-#endif
-  abort();
+  caml_abort(); // defined in misc.h
+}
+
+CAMLnoret void caml_debug_abort(char_os * file_os, int line) {
+  caml_print_loc(file_os, line);
+  fprintf(stderr, "Abort\n");
+  fflush(stderr);
```

Comment:
`stderr` is never buffered. It does not make sense to `fflush` it as this is a no-op.

Reply:
I thought about this, but the current code does flush stderr, and people online seem to think that this is still a good idea for portability/compatibility reasons. (Eg. [here](https://utcc.utoronto.ca/~cks/space/blog/programming/HowToWriteToStderr), which also suggests flushing `stdout` before printing to stderr, maybe we should also do this to avoid lost output.)

Reply:
Stdout is often block buffered, and often goes to the same place as stderr, which is why it's always a good idea to flush stdout before writing to stderr. I _think_ that (Posix) stderr is never block buffered (unless you `setvbuf` it?) but may be line-buffered, rather than unbuffered (I think on Linux it's always unbuffered); so it may be a good idea to flush it after writing a part-line, and after any write if it may have been `setvbuf`ed.